src/Features/Core/Portable/ReplaceConditionalWithStatements/AbstractReplaceConditionalWithStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,295 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.ReplaceConditionalWithStatements;
+
+internal abstract class AbstractReplaceConditionalWithStatementsCodeRefactoringProvider<
+    TExpressionSyntax,
+    TInvocationExpressionSyntax,
+    TConditionalExpressionSyntax,
+    TStatementSyntax,
+    TReturnStatementSyntax,
+    TExpressionStatementSyntax,
+    TLocalDeclarationStatementSyntax,
+    TArgumentSyntax,
+    TVariableSyntax,
+    TVariableDeclaratorSyntax,
+    TEqualsValueClauseSyntax>
+    : CodeRefactoringProvider
+    where TExpressionSyntax : SyntaxNode
+    where TInvocationExpressionSyntax : TExpressionSyntax
+    where TConditionalExpressionSyntax : TExpressionSyntax
+    where TStatementSyntax : SyntaxNode
+    where TReturnStatementSyntax : TStatementSyntax
+    where TExpressionStatementSyntax : TStatementSyntax
+    where TLocalDeclarationStatementSyntax : TStatementSyntax
+    where TArgumentSyntax : SyntaxNode
+    where TVariableSyntax : SyntaxNode
+    where TVariableDeclaratorSyntax : SyntaxNode
+    where TEqualsValueClauseSyntax : SyntaxNode
+{
+    protected abstract bool HasSingleVariable(TLocalDeclarationStatementSyntax localDeclarationStatement, [NotNullWhen(true)] out TVariableSyntax? variable);
+    protected abstract TLocalDeclarationStatementSyntax GetUpdatedLocalDeclarationStatement(SyntaxGenerator generator, TLocalDeclarationStatementSyntax localDeclarationStatement, ILocalSymbol symbol);
+
+    public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+    {
+        var conditionalExpression = await context.TryGetRelevantNodeAsync<TConditionalExpressionSyntax>().ConfigureAwait(false);
+        if (conditionalExpression is null)
+            return;
+
+        var (document, _, cancellationToken) = context;
+
+        // supports:
+        // 1. a = x ? y : z;
+        // 2. var a = x ? y : z;
+        // 3. return x ? y : z;
+        // 4. Invocation(x ? y : z);
```

Comment:
This could have side effects if `Invocation` has a `Conditional` attribute. For example, `Debug.Assert(x ? IsSomething() : IsAnotherThing());`. Extracting the calls out of `Debug.Assert` will get them invoked in release.

There is already a warning annotation for code actions that have side effects. Should we detect that?

Reply:
i'm personally fine either way.  i don't feel strongly that we need to warn here.  this is a refactoring, so it's ok for it to just help the user speed things up.